bsmetadata/preprocessing_tools/html_parser/filters_and_cleaners.py

```diff
@@ -0,0 +1,459 @@
+        self.attrs_to_keep = attrs_to_keep
+        self.start_parsing_at_tag = start_parsing_at_tag
+        self.convert_br_tag_to_breaking_line = convert_br_tag_to_breaking_line
+
+        self.tags_to_remove_alone = (
+            [
+                TagToRemove(FAKE_TAG_BLOCK),
+                TagToRemove(FAKE_TAG_INLINE),
+                TagToRemove(FAKE_TAG_BASIC),
+            ]
+            if self.tags_to_remove_alone is None
+            else self.tags_to_remove_alone
+            + [
+                TagToRemove(FAKE_TAG_BLOCK),
+                TagToRemove(FAKE_TAG_INLINE),
+                TagToRemove(FAKE_TAG_BASIC),
+            ]
+        )
+
+        self.block_elements = BLOCK_ELEMENTS.copy()
+        if self.convert_br_tag_to_breaking_line:
+            self.block_elements.remove("br")
+            self.tags_to_remove_alone.append(TagToRemove("br"))
+
+        self.consecutive_tag_cleaner = ConsecutiveTagCleaner(
+            block_elements=self.block_elements,
+            consecutive_tags_to_fold=consecutive_tags_to_fold,
+        )
+
+        self.attribute_cleaner = AttributeCleaner(attrs_to_keep=attrs_to_keep)
+        self.tag_filter = TagFilter(
+            txt_max_chr_len_alone=txt_max_chr_len_alone,
+            txt_min_chr_len_alone=txt_min_chr_len_alone,
+            tags_exceptions_alone=tags_exceptions_to_txt_max_min_chr_len_alone,
+            txt_max_chr_len_with_content=txt_max_chr_len_with_content,
+            txt_min_chr_len_with_content=txt_min_chr_len_with_content,
+            tags_exceptions_with_content=tags_exceptions_to_txt_max_min_chr_len_with_content,
+            tags_to_remove_alone=self.tags_to_remove_alone,
+            tags_to_remove_with_content=tags_to_remove_with_content,
+        )
+
+    def apply(self):
+        html_str = self.html_str
+        # Traitement n°1: start the parsing at a special tags (mostly tested with <body>)
+        if self.start_parsing_at_tag is not None:
+            root = fromstring(html_str)
+            find = etree.XPath(f"//{self.start_parsing_at_tag}")
+            new_etree = find(root)[0]
+            html_str = etree.tostring(new_etree, method="html", encoding="UTF-8", pretty_print=False).decode("UTF-8")
+            if not html_str.startswith("<html>"):
+                self.tag_filter.tags_to_remove_alone.update({"html": TagToRemove("html")})
+
+                # need to re-add html tag otherwise the fromstring` do something strange
+                html_str = f"<html>{html_str}</html>"
+
+        # Traitement n°2: [all treatments impacting the chr_idx] we removes sub-trees from the HTML + we minify the html
+        html_str = htmlmin.minify(html_str, remove_comments=True, keep_pre=True)
+
+        new_etree = fromstring(html_str)
+
+        self._clean_etree(new_etree)
+
+        html_str = etree.tostring(new_etree, method="html", encoding="UTF-8", pretty_print=False).decode("UTF-8")
+        html_str = htmlmin.minify(html_str, keep_pre=True)
```

Comment:
I don't quite get why we need to call `minify` twice (but I guess it's not too important that I understand every little detail)

Reply:
This is indeed part of some of my choices as the function between the 2 minification operations `self._clean_etree` does not necessarily produce a minified html. 

It will probably be easier to see this on an example. Let's say that our initial html is :
```
\n    <html>\n    <head>\n    </head>\n    <body>\n    <h1>This is a title</h1>\n    <div>small</div>\n    <div>\n    <div>This is a <div>first</div> sub-div in div</div>\n    <div>This is a <div>second</div> sub-div in div</div>\n    </div>\n    <p>This is a paragraph not in div</p>\n    </body>\n    </html>\n
```
(I took the test example `test_parse_html_nested_example_min_length`). When we minify it the first time it becomes : 
```
<html><body> <h1>This is a title</h1> <div>small</div> <div> <div>This is a <div>first</div> sub-div in div</div> <div>This is a <div>second</div> sub-div in div</div> </div> <p>This is a paragraph not in div</p> </body> </html>'
```

But if now, we want to remove the subparts of the DOM corresponding to div tags if the contained text is longer than 7 characters, then the output of `self._clean_etree` will be : 
```
<html><body> <h1>This is a title</h1> <div>small</div>  <p>This is a paragraph not in div</p> </body> </html>
```

We notice in particular that there are two spaces between `</div>` and `<p>`. If we reapply minify, this double space will be reduced to 1:
```
<html><body> <h1>This is a title</h1> <div>small</div> <p>This is a paragraph not in div</p> </body> </html>
```

Please note, as we discussed a long time ago, the result of `minify` will not be a perfect minification but a "satisfactory" one.

Reply:
Perfect, makes sense!